File: exercicios/projeto-final/skeletons/thecocktaildb/kmp/pratica/androidApp/src/main/kotlin/com/puciec/cocktailkmp/DetailScreen.kt
package com.puciec.cocktailkmp

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import com.puciec.cocktailkmp.data.DrinkDetail
import com.puciec.cocktailkmp.data.TheCocktailDbApi

@Composable
fun DetailScreen(
    api: TheCocktailDbApi,
    favoritesStore: FavoritesStore,
    drinkId: String,
    onBack: () -> Unit,
) {
    var detail by remember { mutableStateOf<DrinkDetail?>(null) }
    var isFavorite by remember { mutableStateOf(favoritesStore.load().contains(drinkId)) }
    var error by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(drinkId) {
        // TODO 2 (feature 2 — detalhe): chamar api.fetchDetail(drinkId) e
        // guardar em `detail`. Tratar erro em `error` (try/catch).
        try {
            detail = api.fetchDetail(drinkId)
        } catch (e: Exception) {
            error = e.message
        }
    }

    Column(
        Modifier
            .fillMaxSize()
            .statusBarsPadding()
            .testTag("detail-screen")
            .padding(16.dp)
    ) {
        Text(
            text = "< Voltar",
            modifier = Modifier
                .testTag("detail-back-button")
                .clickable { onBack() },
        )
        when {
            error != null -> Text(error!!)
            detail == null -> Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                CircularProgressIndicator()
            }

            else -> {
                val d = detail!!
                Row(
                    Modifier
                        .fillMaxWidth()
                        .padding(top = 16.dp),
                    horizontalArrangement = Arrangement.SpaceBetween,
                ) {
                    Text(
                        text = d.strDrink,
                        style = MaterialTheme.typography.headlineMedium,
                        modifier = Modifier.testTag("detail-title"),
                    )
                    Text(
                        text = if (isFavorite) "♥" else "♡",
                        modifier = Modifier
                            .testTag("detail-favorite-button")
                            .clickable {
                                // TODO 5 (feature 5 — favoritos): chamar
                                // favoritesStore.toggle(drinkId) e atualizar
                                // `isFavorite` com o resultado.
                                favoritesStore.toggle(d.idDrink)
                                isFavorite = favoritesStore.load().contains(d.idDrink)
                            },
                        style = MaterialTheme.typography.headlineMedium,
                    )
                }
                Text(d.strCategory ?: "")
                Text("Alcoólico: ${d.strAlcoholic ?: ""}")
                Text("Copo: ${d.strGlass ?: ""}")
            }
        }
    }
}

File: exercicios/projeto-final/skeletons/thecocktaildb/flutter/pratica/lib/screens/detail_screen.dart
import 'package:flutter/material.dart';
import '../api/cocktaildb_api.dart';
import '../models/drink.dart';
import '../storage/favorites_store.dart';

class DetailScreen extends StatefulWidget {
  final String drinkId;

  const DetailScreen({super.key, required this.drinkId});

  @override
  State<DetailScreen> createState() => _DetailScreenState();
}

class _DetailScreenState extends State<DetailScreen> {
  final CocktailDbApi _api = CocktailDbApi();
  final FavoritesStore _favoritesStore = FavoritesStore();

  DrinkDetail? _detail;
  bool _isFavorite = false;
  String? _error;

  @override
  void initState() {
    super.initState();
    _load();
  }

  Future<void> _load() async {
    try {
      final detail = await _api.fetchDetail(widget.drinkId);
      final favorites = await _favoritesStore.load();
      setState(() {
        _detail = detail;
        _isFavorite = favorites.contains(widget.drinkId);
      });
    } catch (error) {
      setState(() {
        _error = error.toString();
      });
    }
  }

  Future<void> _toggleFavorite() async {
    // TODO 5 (feature 5 — favoritos): chamar `_favoritesStore.toggle(widget.drinkId)`
    // e atualizar `_isFavorite` via setState.
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        leading: Semantics(
          identifier: 'detail-back-button',
          child: IconButton(
            icon: const Icon(Icons.arrow_back),
            onPressed: () => Navigator.of(context).pop(),
          ),
        ),
        title: const Text('Detalhe'),
      ),
      body: Semantics(
        identifier: 'detail-screen',
        child: _buildBody(),
      ),
    );
  }

  Widget _buildBody() {
    if (_error != null) {
      return Center(child: Text(_error!));
    }
    final detail = _detail;
    if (detail == null) {
      return const Center(child: CircularProgressIndicator());
    }
    return Padding(
      padding: const EdgeInsets.all(16),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Center(
            child: Image.network(detail.thumbUrl, height: 140),
          ),
          const SizedBox(height: 16),
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Semantics(
                identifier: 'detail-title',
                child: Text(
                  detail.name,
                  style: Theme.of(context).textTheme.headlineMedium,
                ),
              ),
              Semantics(
                identifier: 'detail-favorite-button',
                child: IconButton(
                  icon: Icon(
                    _isFavorite ? Icons.favorite : Icons.favorite_border,
                    color: Colors.redAccent,
                  ),
                  onPressed: _toggleFavorite,
                ),
              ),
            ],
          ),
          const SizedBox(height: 8),
          Chip(label: Text(detail.category)),
          const SizedBox(height: 16),
          Text('Alcoólico: ${detail.alcoholic}'),
          Text('Copo: ${detail.glass}'),
        ],
      ),
    );
  }
}
